### PR TITLE
Refactor code to simplify manifest items and their creation

### PR DESCRIPTION
**Pull request by gsnedders**

### CODE REVIEW

File: tools/manifest/item.py [hunk 1]
```diff
@@ -108,6 +88,10 @@ def id(self):
     def meta_flags(self):
         return set(urlparse(self.url).path.rsplit("/", 1)[1].split(".")[1:-1])
 
+    @property
+    def url_base(self):
+        return "/"
```
Comment (gsnedders) [inline tools/manifest/item.py]: This is blatantly wrong. But also currently doesn't break anything (uh, we should have tests!).

File: tools/manifest/item.py [hunk 2]
```diff
@@ -52,12 +37,7 @@ def id(self):
 
     @property
     def meta_flags(self):
-        return set(self.source_file.meta_flags)
-
-    @property
-    def path(self):
-        """The test path relative to the test_root"""
-        return self.source_file.rel_path
+        return set()
```
Comment (jgraham) [inline tools/manifest/item.py]: Is the belief that this isn't used anywhere? If so can we add a warning or something if we don't want to remove it?

Comment (gsnedders) [reply tools/manifest/item.py]: Yes. AFAICT, it is literally only used by `ManifestItem.https` and nowhere else, and we don't need to deal with that for cases where we don't have a URL. I guess `https` should really move to `URLManifestItem`, and `meta_flags` made private?

File: tools/wptrunner/wptrunner/wpttest.py [hunk 3]
```diff
@@ -149,12 +149,12 @@ def update_metadata(self, metadata=None):
...
         timeout = cls.long_timeout if manifest_item.timeout == "long" else cls.default_timeout
         protocol = "https" if hasattr(manifest_item, "https") and manifest_item.https else "http"
-        return cls(manifest_item.source_file.tests_root,
+        return cls(manifest_file.tests_root,
                    manifest_item.url,
                    inherit_metadata,
                    test_metadata,
                    timeout=timeout,
-                   path=manifest_item.source_file.path,
+                   path=os.path.join(manifest_file.tests_root, manifest_item.path),
```
Comment (jgraham) [inline tools/wptrunner/wptrunner/wpttest.py]: You could have a method on the manifest to return this

Comment (gsnedders) [reply tools/wptrunner/wptrunner/wpttest.py]: That seems a bit weird to me, at least.

File: tools/manifest/item.py (hunk 1 above)
Comment (gsnedders) [reply tools/manifest/item.py]: Hah, seemingly this is actually a constant! (We always set it to `/` and `SourceFile` expands the full URL.) And only used by `.url`, so we can just drop this.

**Comment by gsnedders** [comment]
So, despite this mostly being done as refactoring to do other work for #15137, this has made the single-test case (`time ./wpt run --binary=`which chromium` --headless --no-pause chrome infrastructure/assumptions/html-elements.html`) 20% faster on my desktop computer.

File: tools/wptrunner/wptrunner/testloader.py [hunk 4]
```diff
@@ -504,7 +504,7 @@ def iter_tests(self):
             manifest_items = self.chunker(manifest_items)
 
         for test_type, test_path, tests in manifest_items:
-            manifest_file = manifests_by_url_base[iter(tests).next().url_base]
+            manifest_file = manifests_by_url_base["/"]
```
Comment (gsnedders) [inline tools/wptrunner/wptrunner/testloader.py]: This is obviously wrong. However, given `item.url_base` was previously just the constant `"/"`, this isn't a regression. I presume @jgraham will say I need to fix this code though. )

File: tools/manifest/item.py (hunk 1 above)
Comment (jgraham) [reply tools/manifest/item.py]: Note that it's set by the manifest and doenstream projects (notably gecko and servo) have manifests which set values other than `/` e.g. https://searchfox.org/mozilla-central/source/testing/web-platform/mozilla/meta/MANIFEST.json#1768

Comment (jgraham) [reply tools/manifest/item.py]: (That might not mean this code wasn't useless, just checking you're aware of where the url base is used).

File: tools/wptrunner/wptrunner/testloader.py (hunk 4 above)
Comment (jgraham) [reply tools/wptrunner/wptrunner/testloader.py]: You need to fix this code. Or change the iterator to produce something different like `(manifest, source_file, tests)`.

File: tools/wptrunner/wptrunner/wpttest.py [hunk 5]
```diff
@@ -313,32 +313,28 @@ class TestharnessTest(Test):
...
     def __init__(self, tests_root, url, inherit_metadata, test_metadata,
                  timeout=None, path=None, protocol="http", testdriver=False,
-                 jsshell=False, scripts=None):
+                 jsshell=False):
         Test.__init__(self, tests_root, url, inherit_metadata, test_metadata, timeout,
                       path, protocol)
 
         self.testdriver = testdriver
         self.jsshell = jsshell
-        self.scripts = scripts or []
```
Comment (Ms2ger) [inline tools/wptrunner/wptrunner/wpttest.py]: I do need *some* way to get these; they're used in https://searchfox.org/mozilla-central/rev/cb7faaf6b4ad2528390186f1ce64618dea71031e/js/src/tests/jstests.py#404

File: tools/wptrunner/wptrunner/testloader.py (hunk 4 above)
Comment (gsnedders) [reply tools/wptrunner/wptrunner/testloader.py]: So, we only get this right when loading the JSON, not when actually creating the manifest.

File: tools/wptrunner/wptrunner/wpttest.py (hunk 5 above)
Comment (gsnedders) [reply tools/wptrunner/wptrunner/wpttest.py]: @jgraham told me it was unused! :)

Guess we need to add it to the JSON, then.

Comment (jgraham) [reply tools/wptrunner/wptrunner/wpttest.py]: Well it's not that expensive to read from the file. But if the goal is that we don't read any test file data except metadata when the test file hasn't been updated, then yes.

Comment (gsnedders) [reply tools/wptrunner/wptrunner/wpttest.py]: > But if the goal is that we don't read any test file data except metadata when the test file hasn't been updated, then yes.

We shouldn't need to read any test file data to just run the test.

Comment (gsnedders) [reply tools/wptrunner/wptrunner/wpttest.py]: So I've fixed this by putting it in the manifest… but I fear this means we're now gonna be broken until the MANIFEST is regenerated from scratch given it won't contain this until it's regenerated…

Comment (jgraham) [reply tools/wptrunner/wptrunner/wpttest.py]: We can bump the manifest version number to force it to be regenerated. We should probably also make sure that all the manifest upload things regenerate from scratch every time.

**Comment by gsnedders** [comment]
Can't figure out why the `test_wpttest.py` tests are failing; will need to dig in deeper next week.

File: tools/wptrunner/wptrunner/wpttest.py (hunk 5 above)
Comment (gsnedders) [reply tools/wptrunner/wptrunner/wpttest.py]: Following on from discussion on IRC, I've re-added `tests_root` to items (yay memory consumption!) and made it possible for them to create `SourceFile` objects (albeit behind a private property) to fetch the `scripts` from there if the `_extras` doesn't already contain it.

Comment (gsnedders) [reply tools/wptrunner/wptrunner/wpttest.py]: And, to be clear, when we next bump the MANIFEST version we should eliminate all of that.

**Comment by gsnedders** [comment]
The one remaining failure is in `test_list_tests_missing_manifest`. This appears to be caused by us passing `--tests tools/wpt/tests` into the `wpt run` call in that test, and this hella confuses things. We end up generating a manifest for the git repo root but then plenty of other places aren't aware we're doing this.

**Comment by gsnedders** [comment]
Given @Hexcles mentioned memory consumption of manifest generation in https://github.com/web-platform-tests/wpt/issues/12874#issuecomment-465794369:

On Linux/x86_64, on this branch, manifest generation from scratch consumes max 487MB, clean regeneration 202MB.

`master`, by way of comparison requires max 1408MB and 290MB respectively.

**Comment by Hexcles** [comment]
That is definitely promising! I also tested the memory footprint of manifest before (~1.5G from scratch) and it is indeed a problem for folks running tests on VMs (or Taskcluster workers, from forks of this repo) with <=4G memory.

We should also try to free the memory ASAP. How much memory does the manifest take when we are running the tests (without manifest generation)? I assume it'd take take less than generating the manifest, and if so, it'd be great to make sure the delta is freed when the generation is completed.

**Comment by gsnedders** [comment]
And the commit I've just pushed here makes it go down to max 165MB and 187MB respectively (the delta I presume is caused by it loading the old manifest and therefore having that in memory too).

But this does now mean we aren't holding on to any file beyond the point at which we're directly processing that file.

**Review by jgraham — COMMENTED** [review]
THis looks great. Let me double check it doesn't break gecko before approving.

**Comment by jgraham** [comment]
It seems like this broke gecko by calling git unconditionally https://treeherder.mozilla.org/#/jobs?repo=try&author=james%40hoppipolla.co.uk&collapsedPushes=420652&fromchange=551ebec95f9b6ad9b5ac3d8812ee04b61d9024c3&selectedJob=230362281

**Review by jgraham — CHANGES_REQUESTED** [review]
Need to ensure that `git` isn't required when updating with `--work`.

**Comment by gsnedders** [comment]
> Need to ensure that `git` isn't required when updating with `--work`.

That's fixed now.

File: tools/manifest/item.py [hunk 6]
```diff
@@ -38,39 +22,26 @@ def __new__(cls, name, bases, attrs, **kwargs):
...
-    @property
-    def https(self):
-        flags = self.meta_flags
-        return ("https" in flags or "serviceworker" in flags)
-
     def key(self):
         """A unique identifier for the test"""
         return (self.item_type, self.id)
 
     def meta_key(self):
```
[the remark this answers is not in the document]
Comment (gsnedders) [reply tools/manifest/item.py]: I… don't think so? But I don't really have the confidence here. We represent items as sets in various places which means we rely on the hashes to some extent, at least.

[the remark this answers is not in the document]
Comment (jgraham) [reply tools/manifest/item.py]: Sure, but unless we are (still) using the hash to determine whether a test updated or not then we can probably remove this method. Not for this PR anyway.

**Review by jgraham — APPROVED** [review]

**Comment by foolip** [comment]
@gsnedders I've been running a script to regenerate manifests for all tagged commits locally and compare with what can be downloaded from GitHub releases. Starting with merge_pr_15308 there are a lot that have an added slash at the beginning of paths. I'm not sure why that happened, was that ever the intended behavior before, during or after this change?

**Comment by foolip** [comment]
@gsnedders using "dom/nodes/Document-doctype.html" as a test path that's been around since 2014, it looks like the leading slash was originally not there in the manifest and at some point was added. From which commit should it be there, when doing a full rebuild of the manifest?

**Comment by gsnedders** [comment]
@foolip not sure; it shouldn't be there (and isn't today).

**Comment by foolip** [comment]
@gsnedders alright, I'll work on the assumption that it should never have been there. So far it looks like it's there for some manifests created using an incremental update, but not if you recreate the manifest from scratch on the same commit.

**Pull request merged**